src/usr/share/netcheck/netcheck.py

```diff
@@ -13,410 +13,713 @@
+                '_dns_works: Attempting second DNS query for %s on connection "%s" '
+                'using name server %s.', query_names[1], connection_context['id'],
+                nameservers[1])
+            if self._dns_query(
+                    loop_time, connection_context, nameservers[1], query_names[1]):
+                dns_works = True
+                self.logger.trace(
+                    '_dns_works: Second DNS query on connection "%s" successful.',
+                    connection_context['id'])
+            else:
+                connection_context['activated'] = False
+                self.logger.debug(
+                    '_dns_works: Second DNS query for %s failed on connection "%s" using '
+                    'name server %s. Assuming connection no longer has Internet access.',
+                    query_names[1], connection_context['id'], nameservers[1])
 
-            try:
-                self.logger.debug('check_loop: Check loop iteration starting.')
+        return dns_works
 
-                # Periodically connect to the main backup network because the carrier
-                #   requires this.
-                self._use_backup_network()
+    # TODO: Use something more secure than unauthenticated DNS requests. (issue 5)
+    def _dns_query(self, loop_time, connection_context, nameserver, query_name):
+        """Attempts a DNS query for query_name on 'nameserver' via a connection.
 
-                wired_is_connected = self._check_connection_and_check_dns(
-                    self.config['wired_network_name'])
+        loop_time: The datetime representing when the current program loop began.
+        connection_context: Contains stateful information for the connection being checked.
+          Some connection state information is set in this method.
+        nameserver: The IP address of the name server to use in the query.
+        query_name: The DNS name to query.
+        Returns True if successful, False otherwise.
+        """
+        self.logger.trace('_dns_query: Querying %s for %s on connection "%s".', nameserver,
+                          query_name, connection_context['id'])
+        success = False
 
-                if (wired_is_connected):
-                    self.logger.debug('check_loop: Wired network still connected with '
-                                      'successful DNS check.')
-                    current_network_name = self.config['wired_network_name']
-                else:
-                    self.logger.debug('check_loop: Wired network is not connected.')
+        interface_ip = self.network_helper.get_connection_ip(connection_context['id'])
 
-                    wired_connection_success = self._connect_and_check_dns(
-                        self.config['wired_network_name'])
+        if interface_ip is not None:
+            self.resolver.nameservers = [nameserver]
+            try:
+                self.resolver.query(query_name, source=interface_ip)
```

Comment:
Specifying the source interface seems to have stopped working. It doesn't work with dig either, so I suspect a configuration error on my end or a low level OS problem. I'll have to investigate this tonight.